pkg/client/record/event_test.go

```diff
@@ -56,6 +61,57 @@ func (t *testEventSink) Update(e *api.Event) (*api.Event, error) {
 	return e, nil
 }
 
+// PatchEvent records the event for testing.
+func (t *testEventSink) Patch(e *api.Event, p []byte) (*api.Event, error) {
+	if t.OnPatch != nil {
+		return t.OnPatch(e, p)
+	}
+	return e, nil
+}
+
+type OnCreateFunc func(*api.Event) (*api.Event, error)
+
+func OnCreateFactory(testCache map[string]*api.Event, createEvent chan *api.Event) OnCreateFunc {
+	return func(event *api.Event) (*api.Event, error) {
+		// because the accuracy of time decreases to second-level after time is
+		// encoded, we have to manually shift the timestamp here, so the
+		// timestamp in patch will be different even if the same event occurs in
+		// rapid succession.
+		shift := time.Duration(-10 * time.Second)
+		event.FirstTimestamp = unversioned.NewTime(event.FirstTimestamp.Time.Add(shift))
```

Comment:
I would like to remove the dependency on timing completely. This can lead to flaky tests, even if it is unlikely. I was thinking maybe something like:

```go
func OnCreateFactory(testCache map[string]*api.Event, createEvent chan<- *api.Event, ticker func() time.Time) OnCreateFunc {
	return func(event *api.Event) (*api.Event, error) {
		ts := ticker()
		event.FirstTimestamp = ts
		event.LastTimestamp = ts
		...
	}
}

func OnPatchFactory(testCache map[string]*api.Event, createEvent chan<- *api.Event, ticker func() time.Time) OnPatchFunc {
	return func(event *api.Event) (*api.Event, error) {
		ts := ticker()
		// Modify timestamps appropriately
		...
	}
}

...

func TestEventf(t *testing.T) {
	...
	var tickerTimestamp time.Time
	ticker := func() time.Time {
		tickerTimestamp = tickerTimestamp.Add(time.Second)
		return tickerTimestamp
	}
	testEvent := testEventSink{
		OnCreate: OnCreateFactory(testCache, createEvent, ticker),
		OnPatch:  OnPatchFactory(testCache, patchEvent, ticker),
		...
}
```

Would something like this work? Is the only place these timestamp matter in `validateEvent`?